File: scraper/scrapers/arbeitnow.py
"""Arbeitnow API scraper — free, no API key needed. European + remote jobs."""

from __future__ import annotations

from typing import Optional

from rich.console import Console

from .base import BaseScraper, JobResult, USER_AGENT


class ArbeitnowScraper(BaseScraper):
    """Scrape jobs from the Arbeitnow public API."""

    name = "arbeitnow"
    requires_browser = False

    MAX_PAGES = 3

    def __init__(self, console: Console, config: Optional[dict] = None) -> None:
        super().__init__(console, config)

    def scrape(
        self,
        keywords: list[str],
        location: str,
        remote: bool = False,
    ) -> list[JobResult]:
        query = " ".join(keywords)
        self.console.log(
            f"[bold bright_yellow]Arbeitnow[/] Searching for [cyan]'{query}'[/]"
            f" in [cyan]{location or '(any)'}[/]"
        )

        results: list[JobResult] = []
        headers = {"User-Agent": USER_AGENT, "Accept": "application/json"}

        for page_num in range(1, self.MAX_PAGES + 1):
            self.console.log(
                f"[bright_yellow]Arbeitnow[/] Fetching page {page_num}/{self.MAX_PAGES}..."
            )

            try:
                resp = self.http.get(
                    f"https://arbeitnow.com/api/job-board-api?page={page_num}",
                    headers=headers,
                    timeout=15,
                )
                resp.raise_for_status()
                data = resp.json()
            except Exception as exc:
                self.console.log(f"[red]Arbeitnow[/] API request failed: {exc}")
                break

            items = data.get("data", [])
            if not items:
                self.console.log("[bright_yellow]Arbeitnow[/] No more results.")
                break

            query_lower = query.lower()
            query_tokens = set(query_lower.split())

            for item in items:
                try:
                    title = item.get("title", "").strip()
                    if not title:
                        continue

                    # Client-side keyword filtering since API doesn't support search
                    title_lower = title.lower()
                    tags_str = " ".join(item.get("tags", [])).lower()
                    desc_lower = (item.get("description", "") or "").lower()
                    searchable = f"{title_lower} {tags_str} {desc_lower}"

                    # Check if any keyword token appears in title, tags, or description
                    if not any(token in searchable for token in query_tokens):
                        continue

                    company = item.get("company_name", "").strip() or None
                    job_location = item.get("location", "").strip() or None
                    is_remote = item.get("remote", False)
                    if is_remote and job_location:
                        job_location = f"{job_location} (Remote)"
                    elif is_remote:
                        job_location = "Remote"

                    job_url = item.get("url", "")
                    if not job_url:
                        slug = item.get("slug", "")
                        if slug:
                            job_url = f"https://arbeitnow.com/view/{slug}"
                        else:
                            continue

                    # Description — strip HTML
                    description = item.get("description", "")
                    if description:
                        import re
                        description = re.sub(r"<[^>]+>", " ", description)
                        description = re.sub(r"\s+", " ", description).strip()
                        if len(description) > 500:
                            description = description[:500] + "..."

                    created = item.get("created_at") or None
                    if created and isinstance(created, (int, float)):
                        from datetime import datetime
                        created = datetime.fromtimestamp(created).strftime("%Y-%m-%d")

                    tags = item.get("tags", [])

                    results.append(
                        JobResult(
                            title=title,
                            company=company,
                            location=job_location,
                            url=job_url,
                            source=self.name,
                            description=description or None,
                            posted_date=created,
                        )
                    )
                except Exception:
                    continue

            self.console.log(
                f"[bright_yellow]Arbeitnow[/] Found {len(results)} jobs so far."
            )

            # Check if there are more pages
            links = data.get("links", {})
            if not links.get("next"):
                break

            self.rate_limit(0.5, 1.5)

        self.console.log(
            f"[bold bright_yellow]Arbeitnow[/] Finished — {len(results)} jobs collected."
        )
        self.jobs = results
        return results

File: web/src/lib/scrapers/arbeitnow.ts
import type { ScrapeParams, ScrapeResult } from './types';
import { fetchJson, stripHtml, matchesKeywords } from './utils';

interface ArbeitnowJob {
  title?: string; company_name?: string; location?: string;
  remote?: boolean; url?: string; slug?: string;
  description?: string; created_at?: number | string;
  tags?: string[];
}

export async function scrapeArbeitnow(params: ScrapeParams): Promise<ScrapeResult> {
  const _query = params.keywords.join(' ');
  const jobs = [];

  for (let page = 1; page <= 3; page++) {
    const data = await fetchJson<{ data?: ArbeitnowJob[]; links?: { next?: string } }>(
      `https://arbeitnow.com/api/job-board-api?page=${page}`
    );
    const items = data?.data ?? [];
    if (!items.length) break;

    for (const item of items) {
      const title = item.title?.trim();
      if (!title) continue;

      const searchable = `${title} ${(item.tags ?? []).join(' ')} ${item.description ?? ''}`;
      if (!matchesKeywords(searchable, params.keywords)) continue;

      let location = item.location?.trim() || undefined;
      if (item.remote && location) location = `${location} (Remote)`;
      else if (item.remote) location = 'Remote';

      const url = item.url || (item.slug ? `https://arbeitnow.com/view/${item.slug}` : undefined);
      if (!url) continue;

      let posted_date: string | undefined;
      if (typeof item.created_at === 'number') {
        posted_date = new Date(item.created_at * 1000).toISOString().split('T')[0];
      } else if (typeof item.created_at === 'string') {
        posted_date = item.created_at;
      }

      jobs.push({
        title, company: item.company_name?.trim() || undefined,
        location, url, source: 'arbeitnow' as const,
        description: item.description ? stripHtml(item.description) : undefined,
        posted_date,
        skills: item.tags?.join(', ') || undefined,
      });
    }
    if (!data?.links?.next) break;
  }

  return { source: 'arbeitnow', jobs };
}
